ext/opentelemetry-ext-opentracing-shim/src/opentelemetry/ext/opentracingshim/__init__.py

```diff
@@ -0,0 +1,232 @@
+            end_time = util.time_seconds_to_ns(finish_time)
+        self._otel_span.end(end_time=end_time)
+
+    def set_tag(self, key, value):
+        self._otel_span.set_attribute(key, value)
+
+        # Return self for call chaining.
+        return self
+
+    def log_kv(self, key_values, timestamp=None):
+        if timestamp is not None:
+            event_timestamp = util.time_seconds_to_ns(timestamp)
+        else:
+            event_timestamp = util.time_ns()
+
+        event_name = util.event_name_from_kv(key_values)
+        event = Event(event_name, event_timestamp, key_values)
+        self._otel_span.add_lazy_event(event)
+
+        # Return self for call chaining.
+        return self
+
+    def set_baggage_item(self, key, value):
+        raise NotImplementedError
+        # TODO: Implement.
+
+    def get_baggage_item(self, key):
+        raise NotImplementedError
+        # TODO: Implement.
+
+    # TODO: Verify calls to deprecated methods `log_event()` and `log()` on
+    # base class work properly (it's probably fine because both methods call
+    # `log_kv()`).
+
+
+class ScopeWrapper(opentracing.Scope):
+    def close(self):
+        self._span.finish()
+        # TODO: Set active span on OpenTelemetry tracer.
+        # https://github.com/open-telemetry/opentelemetry-python/issues/161#issuecomment-534136274
+
+
+class ScopeManagerWrapper(opentracing.ScopeManager):
+    def __init__(self, tracer):
+        # The only thing the `__init__()` method on the base class does is
+        # initialize `self._noop_span` and `self._noop_scope` with no-op
+        # objects. Therefore, it doesn't seem useful to call it.
+        # pylint: disable=super-init-not-called
+        self._tracer = tracer
+
+    @contextmanager
+    def activate(self, span, finish_on_close):
+        with self._tracer.unwrap().use_span(
+            span.unwrap(), end_on_exit=finish_on_close
+        ) as otel_span:
+            wrapped_span = SpanWrapper(
+                self._tracer, otel_span.get_context(), otel_span
+            )
+            yield ScopeWrapper(self, wrapped_span)
+
+    @property
+    def active(self):
+        span = self._tracer.unwrap().get_current_span()
+        if span is None:
+            return None
+
+        wrapped_span = SpanWrapper(self._tracer, span.get_context(), span)
+        return ScopeWrapper(self, wrapped_span)
+        # TODO: Return a saved instance of SpanWrapper instead of constructing
+        # a new object (and the same for ScopeWrapper?).
+        # https://github.com/open-telemetry/opentelemetry-python/issues/161#issuecomment-534136274
+
+
+class TracerWrapper(opentracing.Tracer):
+    def __init__(self, tracer, scope_manager=None):
+        # If a scope manager isn't provided by the user, create a
+        # `ScopeManagerWrapper` instance and use it to initialize the
+        # `TracerWrapper`.
+        if scope_manager is None:
+            scope_manager = ScopeManagerWrapper(self)
```

Comment:
In the OT Shim Java I'm passing around the OTel Tracer + DistributedContext to avoid the circular reference. Maybe something like that could be done here as well.

Reply:
The problem is that `ScopeManagerWrapper` creates `ScopeWrapper` objects which depend on `SpanWrapper` which has a `SpanWrapper.tracer` property:

>Provides access to the :class:`Tracer` that created this :class:`Span`.

I guess if we can get a reference to an existing `SpanWrapper` object rather than constructing one ad-hoc inside `ScopeManagerWrapper`, this would also solve the circular dependency. This may be related to https://github.com/open-telemetry/opentelemetry-python/issues/161#issuecomment-534136274.

Reply:
>In the OT Shim Java I'm passing around the OTel Tracer + DistributedContext to avoid the circular reference. Maybe something like that could be done here as well.

Looking at the Java OT shim, this problem doesn't exist there because it uses the **OpenTelemetry** tracer rather than the shim as the value for the `tracer` attribute when constructing spans:

https://github.com/open-telemetry/opentelemetry-java/blob/39a0664dbf8bda82c3fe05176b3a87526df2786b/opentracing_shim/src/main/java/io/opentelemetry/opentracingshim/ScopeManagerShim.java#L41
https://github.com/open-telemetry/opentelemetry-java/blob/39a0664dbf8bda82c3fe05176b3a87526df2786b/opentracing_shim/src/main/java/io/opentelemetry/opentracingshim/TelemetryInfo.java#L30

Doing the same would definitely help me get rid of the circular reference *and* get rid of the `tracer` property on `ScopeManagerShim` (see https://github.com/open-telemetry/opentelemetry-python/pull/211#discussion_r334731998). However, is it OK to store an OTel reference in the span's `tracer` attribute which is exposed in the OT API?

https://github.com/opentracing/opentracing-python/blob/7d2e62b5999d3cdec9a9c2720a34a1d11f425461/opentracing/span.py#L88-L96